[IMP] l10n_es_aeat_mod347: Optimización en cálculo, mejoras en el manejo de muchas facturas y automatismo para asignar trimestres a periodos

pablocm-aserti opened the pull request.

pablocm-aserti committed:
[IMP] l10n_es_aeat_mod347: Asignación automática del trimestre al crear periodos y menú para asignarlos a posteriori

pablocm-aserti committed:
[FIX] l10n_es_aeat_mod347: Eliminado almacenamiento de los campos funcionales de totales por trimestre. Durante el cálculo y creación de los registros de empresa, cada factura disparaba el ćalculo de los campos funcionales, lo que hacía inviable calcular el modelo en escenarios con gran cantidad de facturas por empresa.
[IMP] l10n_es_aeat_mod347: Añadido botón para acceder a la vista lista de los registros de empresa y hacer manejable su búsqueda
[IMP] l10n_es_aeat_mod347: Traducciones

pedrobaeza commented on l10n_es_aeat_mod347/account_period_view.xml:
```diff
@@ -25,5 +25,25 @@
+            <field name="code">self.assign_quarter(cr, uid, context['active_ids'], context)</field>
+        </record>
+
+        <record model="ir.values" id="assign_quarter_act_menu">
```
¿Por qué pones esto como acción? Si ahora mismo se hace automático en la creación. Si lo que quieres es para anteriores, puedes proveer un script de migración que lo haga.

pablocm-aserti replied:
Para periodos que ya existan, cuando se instala por primera vez el módulo el script de migración no se llama (o no he conseguido que se llame). ¿Hay algún modo de inicializar los valores de esa columna teniendo acceso a las filas durante la instalación? Un default en la columna no sirve.

pedrobaeza replied:
Hicieron un cambio para que los scripts de migración no se ejecutaran en la primera instalación (que era cuando le pasaba version=None), pero creo que fue en la v8. De todas formas, ¿esto lo ejecuta al instalar, al actualizar, siempre?

pablocm-aserti replied:
Sólo al tener instalada una versión < 1.2, que es a la que estoy subiendo el módulo, y actualizarlo. Si dices que en la 7 actual se dispara al instalar desde 0 puede que sea algún bug, porque estoy probando con openerp-7.0-20130806-231058

pedrobaeza commented on l10n_es_aeat_mod347/mod347.py:
```diff
@@ -25,6 +25,17 @@ class L10nEsAeatMod347Report(orm.Model):
     _description = "AEAT 347 Report"
 
+    def btn_list_records(self, cr, uid, ids, context=None):
+        value = {
```
¿Para qué es esto?

pablocm-aserti replied:
Es el método del botón para abrir la vista tree de los registros de empresa.

pedrobaeza replied:
¿Y para qué? Si tienes ya el one2many incluido en la pestaña. No ofrece demasiadas ventajas, porque cosas como la búsqueda puedes utilizar la del propio navegador (Ctrl+F).

pablocm-aserti replied:
Porque en el caso al que nos hemos enfrentado salen 342 registros y aunque se puede poner "ilimitado" en la paginación y buscar como apuntas, también quería el cliente separar por clave. Sabiendo que el modelo no se debe cambiar para añadir pestañas de emitidas y recibidas como en el 340, la solución menos intrusiva es permitirle usar la caja de búsqueda, que sobre el Ctrl+F ofrece agrupamiento y filtrado.
Entiendo que tal cantidad de registros no es lo habitual y por eso antes de hacer el PR preguntaba si los cambios os parecían convenientes. Si no es el caso no hay problema, nos hacemos un módulo auto-install aparte que añada esta funcionalidad.

pedrobaeza replied:
Sí, la verdad es que lo que comentas es un añadido que puede ser útil en estos casos. Creo que podemos dejarlo puesto que no molesta nada, pero renómbralo a "Open list in new window", que expresa mejor lo que hace.

pedrobaeza commented:
Tal como lo dejas ahora mismo, hay una sucesión que puede romper la integridad, añadiendo un registro hijo, pero no escribiendo nada en el padre. Hay una mejor forma para esto, y es inhibiendo mientras se rellena el informe, los cálculos automáticos de las funciones, y al final habilitarlos y calcular. Eso se hacía con un parámetro por contexto que se llamaba defer_computation o algo así. Echa un vistazo, y de esa manera obtenemos el mismo resultado, pero sin comprometer esa integridad.

pablocm-aserti commented:
Gracias Pedro, no conocía la existencia de ese parámetro, lo investigaré.
Lo que me queda es la duda de qué es lo que rompe la integridad, si lo único que se hace es eliminar el store. Un store mal definido sí entiendo que se le pueden escapar condiciones en las que debe actualizar el dato, pero si lo haces funcional puro cada vez que se requiere el campo se calcula, ¿no?

pedrobaeza commented:
Efectivamente, ahora que lo veo otra vez, el dato no se almacena (que es lo que supongo llamas funcional puro). En ese caso, no hace falta hacer nada de lo del defer.

pedrobaeza commented on l10n_es_aeat_mod347/account_period_view.xml:
```diff
@@ -25,5 +25,25 @@
             </field>
         </record>
 
+		<record id="act_assign_quarter" model="ir.actions.server">
```
Indentación incorrecta

pedrobaeza commented on l10n_es_aeat_mod347/i18n/es.po:
```diff
@@ -631,7 +586,7 @@ msgstr "Procesada"
 #: field:l10n.es.aeat.mod347.report,representative_vat:0
 msgid "L.R. VAT number"
-msgstr "NIF R.L."
+msgstr "CIF/NIF R.L."
```
CIF ya no existe desde hace mucho. No lo cambies

pablocm-aserti replied:
Muchas de las traducciones han sido cambios involuntarios que han surgido al exportar las traducciones que había cargadas. Por favor, no te molestes en revisarlos porque voy a ver si puedo evitarlos.
Disculpa todo el trabajo que te estoy dando

pedrobaeza replied:
OK, espero entonces a que me avises

pablocm-aserti committed:
[MOD] l10n_es_aeat_mod347: Sustituido el action multi para asignar trimestre por script de migración
[IMP] l10n_es_aeat_mod347: Sólo se crean registros de empresa para las nacionales o cuyo país no esté definido
[FIX] l10n_es_aeat_mod347: Cambiado string del botón para abrir la lista de registros de empresa
[FIX] l10n_es_aeat_mod347: Traducciones exportadas en una bbdd limpia

pablocm-aserti commented:
Ya están los cambios que me has comentado: se ha quitado el action multi, se ha añadido en su lugar el script de migración y se ha cambiado el string del botón.
Adicionalmente he puesto una comprobación para que no tenga en cuenta partners no nacionales, porque salían todos salvo que se excluyan a mano uno a uno, lo que no es práctico

El problema del script de migración que no se lanza al instalar por primera vez sigue ocurriendo en la última versión de la 7

pedrobaeza commented on l10n_es_aeat_mod347/mod347.py:
```diff
@@ -278,6 +290,12 @@ def calculate(self, cr, uid, ids, context=None):
+                                                        context=context)
+                    # Only national
+                    # Undefined are left so that they can be corrected
+                    if address.country_id and address.country_id.code != 'ES':
```
Esto es lo único que no me convence. Ya se tiene una casilla para excluir a algún partner del 347. Se puede poner para que esa casilla se marque o desmarque al cambiar el país, pero no poner aquí este criterio oscuro que no tiene reflejo en el interfaz.

pablocm-aserti replied:
Esa casilla está bien cuando se tiene alguna excepción para algún nacional, pero si por norma los extranjeros no se declaran en el 347, lo que se debe esperar del modelo es que no los tenga en cuenta de base ¿no?
Parece un fallo que los incluya cuando nunca deben salir y una molestia tener que ir uno por uno para corregirlo manualmente.

pedrobaeza replied:
Me parece bien el cambio si:
- Por un lado, lo documentas en README y en la propia pantalla del modelo (con un <label> o <p> en ella bastaría).
- Por otro, ocultas la casilla en el partner si el país es distinto de España.

pablocm-aserti committed:
[FIX] l10n_es_aeat_mod347: Revertida la modificación que dejaba fuera de la declaración a los extranjeros. Las operaciones no-intracomunitarias de servicios sí se declaran.

pablocm-aserti committed:
[FIX] l10n_es_aeat_mod347: Script de migración que faltó por añadir en el commit ba60356

pablocm-aserti committed:
Merge branch '7.0' of git://github.com/OCA/l10n-spain into OCA-7.0

Conflicts:
	l10n_es_aeat_mod347/i18n/es.po
	l10n_es_aeat_mod347/mod347.py

pablocm-aserti committed:
Merge branch 'OCA-7.0' into 7.0

pablocm-aserti committed:
[FIX] l10n_es_aeat_mod347: PEP8

pablocm-aserti committed:
[FIX] l10n_es_aeat_mod347: Traducción

pablocm-aserti commented:
He revertido la exclusión de las empresas extranjeras, porque si se facturan servicios a no intracomunitarias [sí deben declararse](https://www5.aeat.es/es13/s/iafriafrc05f/) y también he añadido el script de migración, que se me pasó en el anterior commit. En teoría no queda nada pendiente.

pedrobaeza commented:
Muchas gracias, Pablo. Mergeo esto, y ahora reviso lo que me comentas del 349.
